File: tests/gf128_test.py
from specs.gf128 import *
from sys import exit
import json

def main(x: int) -> None:
    file = open('tests/test_vectors/gf128_test_vectors.json')
    gf128_test_vectors = json.load(file)
    for i in range(len(gf128_test_vectors)):
        msg = bytes.from_hex(gf128_test_vectors[i]['input'])
        k   = bytes.from_hex(gf128_test_vectors[i]['key'])
        expected = bytes.from_hex(gf128_test_vectors[i]['output'])
        computed = gmac(msg,k)
        if (computed == expected):
            print("GF128 Test ",i," passed.")
        else:
            print("GF128 Test ",i," failed:")
            print("expected mac:",expected)
            print("computed mac:",computed)
            exit(1)


if __name__ == "__main__":
    main(0)

File: compiler/fstar-compiler/specs/tests/gf128_test.ml
open Yojson.Basic.Util
open Testutil

let main() =
  let json = Yojson.Basic.from_file "../../../tests/test_vectors/gf128_test_vectors.json" in
  let jsonl = Yojson.Basic.Util.to_list json in
  List.iteri (fun i j ->
      let msg = read_bytes j "input" in
      let key = read_bytes j "key" in
      let output = read_bytes j "output" in
      let computed = Gf128.gmac msg key in
      if output = computed then Printf.printf "GF128 Test %d passed.\n" i
      else (Printf.printf "GF128 Test %d failed.\n" i;
            Printf.printf "expected tag: %s\n" (hex_of_bytes output);
            Printf.printf "computed tag: %s\n" (hex_of_bytes computed));
      ()) jsonl;
  ()

let _ = main()
